### PR TITLE
Refactor selection box layer - Pass 1

### PR DESCRIPTION
**Pull request by aicioara**
Part of #2623 

I will be doing 2 passes because I will get more context (and better sense of best practices) by refactoring other files. However, please flag anything that seems unusual.

As part of this: 
- **Test semantics: hierarchy, naming, testing, but no coverage check**
- Applied best practices from the wiki page
- Factored out common code into `beforeEach()` clauses
- No nested `beforeEach()`es
- Remade hierarchy as shown bellow
screen shot 2015-08-24 at 12 59 14 pm

### CODE REVIEW

File: test/components/selectionBoxLayerTests.ts [hunk 1]
```diff
@@ -1,193 +1,244 @@
...
+        sbl.renderTo(svg);
+        assert.strictEqual(sbl.boxVisible(), false, "The box is not visible by default after rendering to svg");
+
+        assert.strictEqual(sbl.boxVisible(true), sbl, "Setting the boxVisible attribute returns the selection box layer");
+        assert.strictEqual(sbl.boxVisible(), true, "Setting the boxVisible to true attribute works");
+        sbl.boxVisible(false);
+        assert.strictEqual(sbl.boxVisible(), false, "Setting the boxVisible attribute to false works");
+
+        svg.remove();
+      });
```
Comment (bluong) [inline test/components/selectionBoxLayerTests.ts]: This crosses a very fine line of testing, but I have thinking about how to effectively test getters... only to end with the conclusion of not to test them directly (The best way to test them directly is to pull the corresponding variable backing the getter but then this leads into all sorts of trouble).

The best thing I can think of is to utilize the getters as our means of detecting correctness for the rest of the code (If we use a setter, then the getter should definitely reflect the set value) (If a state-modifying function is called, the getter should definitely reflect the modified state).

Sure we can use these unit tests to prove that the function exists, but I am unsure of its usefulness beyond that (Also this specific idea would be not useful for typescript consumers)

File: test/components/selectionBoxLayerTests.ts [hunk 2]
```diff
@@ -1,193 +1,244 @@
...
+
+        sbl.renderTo(svg);
+        assert.strictEqual(sbl.boxVisible(), false, "The box is not visible by default after rendering to svg");
+
+        assert.strictEqual(sbl.boxVisible(true), sbl, "Setting the boxVisible attribute returns the selection box layer");
+        assert.strictEqual(sbl.boxVisible(), true, "Setting the boxVisible to true attribute works");
+        sbl.boxVisible(false);
+        assert.strictEqual(sbl.boxVisible(), false, "Setting the boxVisible attribute to false works");
+
+        svg.remove();
```
Comment (bluong) [inline test/components/selectionBoxLayerTests.ts]: Suggestion - if we create a component, we destroy the component before removal.

File: test/components/selectionBoxLayerTests.ts [hunk 3]
```diff
@@ -1,193 +1,244 @@
...
+      });
+
+      it("does not error on destroy() when scales are not inputted", () => {
+        sbl.renderTo(svg);
+        assert.doesNotThrow(() => sbl.destroy(), Error, "can destroy even with no scales");
+
+        svg.remove();
+      });
```
Comment (bluong) [inline test/components/selectionBoxLayerTests.ts]: I really don't think this test should exist but should more be embedded into our unit tests or tested elsewhere?

File: test/components/selectionBoxLayerTests.ts [hunk 4]
```diff
@@ -1,193 +1,244 @@
...
+
+      it("can get and set the bounds() property", () => {
```
Comment (bluong) [inline test/components/selectionBoxLayerTests.ts]: Same description as above about getters

**Comment by bluong** [comment]
A few suggestions

File: test/components/selectionBoxLayerTests.ts (hunk 1 above)
Comment (aicioara) [inline test/components/selectionBoxLayerTests.ts]: As @jtlan was saying. We have to build up our tests on axioms. This tests proves that the getter and the setter are having the expected behavior, so we can use them later assuming they are correct. 

I must agree that the usefulness of this test is below the average though.

File: test/components/selectionBoxLayerTests.ts (hunk 2 above)
Comment (aicioara) [inline test/components/selectionBoxLayerTests.ts]: I like that

File: test/components/selectionBoxLayerTests.ts (hunk 4 above)
Comment (aicioara) [inline test/components/selectionBoxLayerTests.ts]: ?

**Comment by aicioara** [comment]
I liked the idea of destroying the components, hence I applied it. Not sure what you mean about the getters / setters

**Comment by bluong** [comment]
I think having tests to test the setters is perfectly valid, but I do not think it makes sense for us to explicitly say that we are testing the getters.

Thus, in this scenario, I would say to change the test names to just saying that you can use the setters and then inside the test just use the getters to get at the set values.

Something like
`it("can set if the box is visible")`

**Comment by aicioara** [comment]
Sounds good

**Pull request merged**